source/fuzz/fuzzer_pass_add_composite_extract.cpp

```diff
@@ -0,0 +1,150 @@
+  std::vector<TransformationCompositeExtract> transformations;
+
+  ForEachInstructionWithInstructionDescriptor(
+      [this, &composite_synonyms, &transformations](
+          opt::Function* function, opt::BasicBlock* block,
+          opt::BasicBlock::iterator inst_it,
+          const protobufs::InstructionDescriptor& instruction_descriptor) {
+        if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(SpvOpCompositeExtract,
+                                                          inst_it)) {
+          return;
+        }
+
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfAddingCompositeExtract())) {
+          return;
+        }
+
+        auto available_composites = FindAvailableInstructions(
+            function, block, inst_it,
+            [](opt::IRContext* ir_context, opt::Instruction* inst) {
+              return inst->type_id() && inst->result_id() &&
+                     fuzzerutil::IsCompositeType(
+                         ir_context->get_type_mgr()->GetType(inst->type_id()));
+            });
+
+        std::vector<const protobufs::DataDescriptor*> available_synonyms;
+        for (const auto* dd : composite_synonyms) {
+          if (fuzzerutil::IdIsAvailableBeforeInstruction(
+                  GetIRContext(), &*inst_it, dd->object())) {
+            available_synonyms.push_back(dd);
+          }
+        }
+
+        if (available_synonyms.empty() && available_composites.empty()) {
+          return;
+        }
+
+        uint32_t composite_id = 0;
+        std::vector<uint32_t> indices;
+
+        if (available_synonyms.empty() || (!available_composites.empty() &&
+                                           GetFuzzerContext()->ChooseEven())) {
+          const auto* inst =
+              available_composites[GetFuzzerContext()->RandomIndex(
+                  available_composites)];
+          composite_id = inst->result_id();
+
+          const auto* type_inst =
+              GetIRContext()->get_def_use_mgr()->GetDef(inst->type_id());
+          assert(type_inst && "Composite instruction has invalid type id");
+
+          uint32_t number_of_members = 0;
+          switch (type_inst->opcode()) {
+            case SpvOpTypeArray:
+              number_of_members =
+                  fuzzerutil::GetArraySize(*type_inst, GetIRContext());
+              break;
+            case SpvOpTypeVector:
+            case SpvOpTypeMatrix:
+              number_of_members = type_inst->GetSingleWordInOperand(1);
+              break;
+            case SpvOpTypeStruct:
+              number_of_members = type_inst->NumInOperands();
+              break;
+            default:
+              assert(false && "|inst| is not a composite");
+              break;
+          }
+
+          if (number_of_members == 0) {
+            return;
+          }
+
+          indices = {GetFuzzerContext()->GetRandomCompositeExtractIndex(
+              number_of_members)};
+        } else {
```

Comment:
If the composite is multi-level, e.g. an array of structs containing matrix fields, it would be pretty cool if this pass could, with some probability, descend deeper into the structure to produce a chain of indices.  How hard would that be to do?

What I mean is, e.g. if we had:

```
%10 = OpTypeVector %float 3
%11 = OpTypeMatrix %10 4
%12 = OpTypeStruct %10 %11 %11
%13 = OpTypeArray %12 %int_100
%14 = ... something of type %13 ...
```

then we might end up doing:

```
%100 = OpCompositeExtract %13 %14 2
```

or

```
%100 = OpCompositeExtract %11 %14 2 1 1
```

or

```
%100 = OpCompositeExtract %10 %14 2 1 1 0
```

Reply:
Fixed in 4514ef2.